### PR TITLE
7425 appliance plex

### LINKED ISSUE #7425: [Core appliances pages] Build page: /appliance/plex/

### PR DESCRIPTION
**Pull request by deadlight**
## Done

- Created appliance/plex page

[copy](https://docs.google.com/document/d/1zgOl4GCK6O6Clv6oxp-DXDSKYIKO2pvpXaYXALiTlpo/edit) | [design](https://app.zeplin.io/project/5eb3d16a91c4117789940141/screen/5eb3fd0a59a84577c29a34ff) | [journey docs](https://docs.google.com/presentation/d/1v5O3iZ33FSrlsQwhESGp3XCvbfNqYLsLvymFxbQcIss/edit#slide=id.g84f83eaa6c_0_225)

## QA

- Check out this feature branch
- Run the site using the command `./run serve` or `dotrun`
- View the site locally in your web browser at: http://0.0.0.0:8001/appliance/plex
- Change the select box and see that the download button URL changes (the pages don't exist yet)
- Run through the following [QA steps](https://canonical-web-and-design.github.io/practices/workflow/qa-steps.html)

## Issue / Card

Fixes #7425

### CODE REVIEW

File: static/js/src/appliance.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+}
+
+appliance.downloadFormsInit = function(downloadForms) {
+  if (downloadForms) {
+    downloadForms.forEach(function(container) {
+      const linkElement = container.querySelector('.js-download-link');
+      const selectElement = container.querySelector('.js-platform-select');
+      if (linkElement && selectElement) {
+        selectElement.addEventListener('change', function() {
+          appliance.updateURL(linkElement, selectElement);
```
Comment (anthonydillon) [inline static/js/src/appliance.js:18]: Why not post the option as a form to the thank you page and conditionally download and select tab based on the option?

File: static/js/src/appliance.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        });
+      }
+    });
+  }
+}
+
+appliance.downloadFormsInit(document.querySelectorAll(".js-download"));
```
Comment (anthonydillon) [inline static/js/src/appliance.js:25]: We shouldn't require js to download something.

File: templates/appliance/plex.html [hunk 3]
```diff
@@ -0,0 +1,171 @@
+{% extends "appliance/_base-appliance.html" %}
+
+{% block title %}Plex Media Server Ubuntu Appliance{% endblock %}
```
Comment (anthonydillon) [inline templates/appliance/plex.html:3]: Probably should separate "Plex Media Server - Ubuntu Appliance"

File: templates/appliance/plex.html [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+{% block meta_description %}{% endblock %}
```
Comment (anthonydillon) [inline templates/appliance/plex.html:5]: Will this leave the description blank? Maybe we should put something here.

File: templates/appliance/plex.html [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    <div class="col-8">
+      <div class="p-media-object u-no-margin--bottom">
+        <img src="https://assets.ubuntu.com/v1/402069bd-plex-media-server.png?w=75" class="p-media-object__image is-round" height="75" width="75">
+        <div class="p-media-object__details">
+          <h1>
+            Plex Media Server
+          </h1>
+          <p class="p-heading--three">Ubuntu Appliance</p>
+          <ul class="p-inline-list--middot">
+            <li class="p-inline-list__item">By Plex, Inc.</li>
```
Comment (anthonydillon) [inline templates/appliance/plex.html:22]: Missing the username in brackets

File: templates/appliance/plex.html [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+        <div class="p-form__group">
+          <label for="platform" class="u-off-screen">Platform</label>
+          <select id="platform" class="js-platform-select">
+            <option value="rpi">Raspberry Pi</option>
+            <option value="kvm">KVM</option>
+            <option value="intel-nuc">Intel NUC</option>
+          </select>
+        </div>
+        <div class="p-form__group">
+          <a href="/appliance/plex/rpi" class="p-button--positive p-form__control js-download-link">Download</a>
```
Comment (anthonydillon) [inline templates/appliance/plex.html:39]: The copy docs say this should like to /appliance/plex/thank-you

File: templates/appliance/plex.html [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+        <li class="p-list__item is-ticked">Get personalized and trusted video news from over 190 global publisher partners (and growing!), including CBS, Financial Times, Euronews, and top local news sources for over 80% of markets in the US</li>
+        <li class="p-list__item is-ticked">Cut the cord! Go premium with a Plex Pass and use Plex Live TV & DVR so you can watch and record free over-the-air HD TV, like NBC, ABC, CBS, and FOX, with any supported digital antenna and tuner</li>
+        <li class="p-list__item is-ticked">Enjoy free features, like remote access, sharing, security, and casting, see your media beautifully enhanced with rich descriptions, artwork, ratings, and more</li>
+        <li class="p-list__item is-ticked">Enjoy premium features with a Plex Pass subscription: DVR & Live TV, Camera Upload, Premium Photos, Mobile Sync, Parental Controls, VIP Perks, and much more</li>
+        <li class="p-list__item is-ticked">Want to use Plex on all of your devices, any time, anywhere? Visit <a href="https://www.plex.tv/">https://www.plex.tv/</a> for more information.</li>
```
Comment (anthonydillon) [inline templates/appliance/plex.html:84]: External link?

File: templates/appliance/plex.html [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+<section class="p-strip--light">
+  <div class="row">
+    <div class="col-8">
+      <h2>Get started</h2>
+      <p>Once you have installed and set up your Plex Media Server Ubuntu appliance you will be taken through set up in your browser. If you have any more questions or want to go further, the developers of Plex Media Server have lots of documentation to help you get started.</p>
+      <a href="https://snapcraft.io/plexmediaserver" class="p-button--neutral p-link--external">Get Started</a>
+    </div>
+    <div class="col-4 u-align--center u-vertically-center">
+      <img src="https://assets.ubuntu.com/v1/402069bd-plex-media-server.png?w=150" height="150" width="150">
```
Comment (anthonydillon) [inline templates/appliance/plex.html:151]: Should this one be w=200 like the others above?

File: templates/appliance/plex.html [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+    </div>
+  </div>
+</section>
+
+<section class="p-strip">
+  <div class="row">
+    <div class="col-8">
+      <h2>Build your own appliance</h2>
+      <p>Making with Ubuntu is easy. If you have an application that you would like to make a part of the Ubuntu appliance portfolio, read our guidelines and get in touch. Soon you will have a page like this.</p>
+      <a href="/appliance/contact-us" class="p-button--neutral">Get in touch</a>
```
Comment (anthonydillon) [inline templates/appliance/plex.html:161]: Missing modal revoke class

**Comment by anthonydillon** [comment]
## QA issues
- the images don't lightbox
- the bottom sections images should be hidden on small screens
- the key has to much bottom margin from the value in the details list

File: static/js/src/appliance.js (hunk 1 above)
Comment (deadlight) [reply static/js/src/appliance.js:18]: The download pages are going to be at different URLs. 

I was going to see how we did with this review and then just add three links as a no-JS fallback.

Thinking about it, a backend form handler doing the redirect might be a better solution.

File: templates/appliance/plex.html (hunk 4 above)
Comment (deadlight) [reply templates/appliance/plex.html:5]: Have requested one in the copydoc

File: templates/appliance/plex.html (hunk 6 above)
Comment (deadlight) [reply templates/appliance/plex.html:39]: There's no link stated in the copydoc. The new flow they introduced yesterday has different pages for each platform's download page

File: templates/appliance/plex.html (hunk 8 above)
Comment (deadlight) [reply templates/appliance/plex.html:151]: Being round, it looked a bit huge at 200. I'll see what design think

File: templates/appliance/plex.html (hunk 9 above)
Comment (deadlight) [reply templates/appliance/plex.html:161]: ah, yes, I forgot that this was already landed

**Comment by deadlight** [comment]
@anthonydillon thanks for the review. re: the lightbox, as discussed yesterday, I'll add that as a separate branch to keep this one smaller.

I'll take a look at the other design tweaks and we can have a chat about the solution to the download flow.

File: static/js/src/appliance.js (hunk 2 above)
Comment (deadlight) [reply static/js/src/appliance.js:25]: Agreed. I was going to add a no-JS fallback but was expecting some discussion around this solution.

File: static/js/src/appliance.js (hunk 1 above)
Comment (anthonydillon) [reply static/js/src/appliance.js:18]: From the copy docs I've looked at that all follow the same structure. The project page and a thank you page. Is there a different one?

File: templates/appliance/plex.html (hunk 4 above)
Comment (anthonydillon) [reply templates/appliance/plex.html:5]: Thanks but for now let's grab a meaningful paragraph from the description

File: templates/appliance/plex.html (hunk 6 above)
Comment (anthonydillon) [reply templates/appliance/plex.html:39]: Did that change from stand up yesterday?

File: templates/appliance/plex.html (hunk 4 above)
Comment (deadlight) [reply templates/appliance/plex.html:5]: Yeah, will do

File: templates/appliance/plex.html (hunk 6 above)
Comment (deadlight) [reply templates/appliance/plex.html:39]: We found out about it at the 10:30 sync yesterday

**Comment by danielmutis** [comment]
@deadlight Can we make the strip at the top shallow, please.

Also, remove the separators here, leaving only the bottom one:
Screenshot 2020-05-12 at 11 57 14

File: templates/appliance/plex.html [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+  <div class="row">
+    <div class="col-8">
+      <div class="p-media-object u-no-margin--bottom">
+        <img src="https://assets.ubuntu.com/v1/402069bd-plex-media-server.png?w=75" class="p-media-object__image is-round" height="75" width="75">
+        <div class="p-media-object__details">
+          <h1>
+            Plex Media Server
+          </h1>
+          <p class="p-heading--three">Ubuntu Appliance</p>
+          <ul class="p-inline-list--middot">
```
Comment (anthonydillon) [inline templates/appliance/plex.html:21]: This needs `u-no-margin--bottom`

File: templates/appliance/plex.html [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+        </div>
+        <div class="p-form__group">
+          <submit href="/appliance/plex/rpi" class="p-button--positive p-form__control">Download</submit>
+        </div>
+      </form>
+    </div>
+  </div>
+</section>
+
+<section class="p-strip">
```
Comment (anthonydillon) [inline templates/appliance/plex.html:46]: This one could do with `u-no-margin--bottom` to bring the description up closer.

File: templates/appliance/plex.html [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+        </li>
+      </ul>
+    </div>
+  </div>
+</section>
+
+<section class="p-strip">
+  <div class="row">
+    <div class="col-8">
+      <h2 class="p-heading--three">Plex magically organizes your media libraries and streams them to any device</h2>
```
Comment (anthonydillon) [inline templates/appliance/plex.html:75]: This should be a heading four

File: templates/appliance/plex.html [hunk 13]
```diff
@@ -0,0 +1,163 @@
+{% extends "appliance/_base-appliance.html" %}
+
+{% block title %}Plex Media Server | Ubuntu Appliance{% endblock %}
+
+{% block meta_description %}Plex magically organizes your media libraries and streams them to any device – including all your video, music, and photo libraries.{% endblock %}
+
+{% block meta_copydoc %}https://docs.google.com/document/d/1zgOl4GCK6O6Clv6oxp-DXDSKYIKO2pvpXaYXALiTlpo/edit#heading=h.c6wjbvvp8mcm{% endblock meta_copydoc %}
```
Comment (anthonydillon) [inline templates/appliance/plex.html:7]: Can you remove the heading link here

**Comment by deadlight** [comment]
@anthonydillon made your suggested changes

**Comment by spencerbygraves** [comment]
@deadlight A couple of changes from me please.

- The title should be h2
- "Ubuntu Appliance" should be h4
- The other titles "Plex magically organizes..." and "Snaps in the image" should also be h4

This will then be the same as snap detail pages

The icon doesn't align nicely with the title, can we do something similar to the snap details pages? Not sure if that has some bespoke styling.

Thanks

**Comment by spencerbygraves** [comment]
@deadlight also, the button with icon looks a bit off.

https://vanillaframework.io/docs/examples/patterns/buttons/icon

**Comment by deadlight** [comment]
@spencerbygraves I updated the headings and a few other things for Ant's review. I'm rebuilding the demos.

Those buttons aren't buttons with icons, they're external links inside a button. They're used all over the site, we can make a change but it should be sitewide rather than just here.

**Comment by deadlight** [comment]
@spencerbygraves if you don't like those buttons with external icons then we can always make them text links. The number of buttons is a little overwhelming.

**Comment by spencerbygraves** [comment]
@deadlight good point, maybe links would be better. If @danielmutis doesn't object.

**Comment by danielmutis** [comment]
@spencerbygraves @deadlight text links work for me 👍🏾

**Comment by deadlight** [comment]
@spencerbygraves @danielmutis updated the demo

File: templates/appliance/plex.html [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+        <div class="p-form__group">
+          <label for="platform" class="u-off-screen">Platform</label>
+          <select id="platform">
+            <option value="rpi">Raspberry Pi</option>
+            <option value="kvm">KVM</option>
+            <option value="intel-nuc">Intel NUC</option>
+          </select>
+        </div>
+        <div class="p-form__group">
+          <submit href="/appliance/plex/rpi" class="p-button--positive p-form__control">Download</submit>
```
Comment (anthonydillon) [inline templates/appliance/plex.html:39]: Never seen this before. Do you mean `<button type="submit"...>`

Comment (deadlight) [reply templates/appliance/plex.html:39]: lol, yes. Karl's brain: "I'll just change that to a submit button"

**Comment by anthonydillon** [comment]
I still don't see any secondary navigation?

**Comment by spencerbygraves** [comment]
@deadlight it was just the spacing above the select on small screen, then all good.

image

**Comment by lyubomir-popov** [comment]
@deadlight @spencerbygraves @danielmutis Some comments - a bit outside of what we discussed, but I think still important:

#### Header

The large media object is not suitable for this use in my opinion. It wasn't designed for a 2-level heading hierarchy, adding an h4 in between creates a lot of white space (which would be fine in a blog post, but not a lockup). 

For comparison: 

Original large media object:
image

This pr:

image

The same snap as it appears on snapcraft:
image

My proposal here (I've tried to keep changes to the absolute minimum):

image

Outside of that, we need to determine optimal icon sizes for each of our headings, and be able to invoke those outside and inside patterns. The vanilla issue for this is [here](https://github.com/canonical-web-and-design/vanilla-framework/issues/3061)

**Comment by lyubomir-popov** [comment]
@deadlight @spencerbygraves @danielmutis 

For the second use of the media object, had to resort to a lot of utilities, but here's my proposal:
image

```
<div class="col-4">
  <h2 class="p-heading--four">Snaps in the image</h2>
  <div class="p-media-object">
    <img src="https://assets.ubuntu.com/v1/402069bd-plex-media-server.png?w=75" class="p-media-object__image is-round" height="75" width="75">
    <div class="p-media-object__details">
      <h3 class="p-heading--5 u-no-margin--bottom u-sv-1">
        <a href="https://snapcraft.io/plexmediaserver" class="p-link--external">Plex Media Server</a>
      </h3>
      <p class="p-muted-text">By Plex, Inc. (plexinc)</p>
      <p class="u-sv-1"><span class="p-muted-text">Channel: </span>latest/stable</p>
      <p class="u-sv-1"><span class="p-muted-text">Version: </span>1.19.3.2764-ef515a800 </p>
      <p><span class="p-muted-text">Published: </span>29 April 2020 </p>
    </div>
  </div>
</div>
```

**Comment by anthonydillon** [comment]
@deadlight I think if you use `persist` instead of `hidden` in the navigation it will display the parent navigation.

File: templates/appliance/plex.html [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+<section class="p-strip is-shallow">
+  <div class="row">
+    <div class="col-8">
+      <div class="p-media-object u-no-margin--bottom">
+        <img src="https://assets.ubuntu.com/v1/402069bd-plex-media-server.png?w=90" class="p-media-object__image is-round" height="90" width="90">
+        <div class="p-media-object__details">
+          <h1 class="p-heading--three u-no-margin--bottom">
+            Plex Media Server
+          </h1>
+          <p class="p-muted-heading"">Ubuntu Appliance</p>
```
Comment (anthonydillon) [inline templates/appliance/plex.html:20]: Double quotes at the end of the class.

File: templates/appliance/plex.html [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+          <ul class="p-inline-list--middot u-no-margin--bottom">
+            <li class="p-inline-list__item">By Plex, Inc. (plexinc)</li>
+            <li class="p-inline-list__item">Entertainment<br class="u-hide--medium "</li>
```
Comment (anthonydillon) [inline templates/appliance/plex.html:23]: Missing be closing bracket

File: templates/appliance/plex.html [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+        <div class="p-form__group">
+          <label for="platform" class="u-off-screen">Platform</label>
+          <select id="platform">
+            <option value="rpi">Raspberry Pi</option>
+            <option value="kvm">KVM</option>
+            <option value="intel-nuc">Intel NUC</option>
+          </select>
+        </div>
+        <div class="p-form__group">
+          <input type="submit" href="/appliance/plex/rpi" class="p-button--positive p-form__control" value="Download">
```
Comment (anthonydillon) [inline templates/appliance/plex.html:39]: Href on a submit button?

File: templates/appliance/plex.html [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+        </div>
+      </div>
+    </div>
+    <div class="col-4 u-align--right">
+      <form class="p-form--inline" action="#">
+        <div class="p-form__group">
+          <label for="platform" class="u-off-screen">Platform</label>
+          <select id="platform">
+            <option value="rpi">Raspberry Pi</option>
+            <option value="kvm">KVM</option>
```
Comment (anthonydillon) [inline templates/appliance/plex.html:34]: This value should be pc

File: templates/appliance/plex/index.html [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+
+{% block meta_copydoc %}https://docs.google.com/document/d/1zgOl4GCK6O6Clv6oxp-DXDSKYIKO2pvpXaYXALiTlpo/edit{% endblock meta_copydoc %}
+
+{% block content %}
+
+<section class="p-strip is-shallow">
+  <div class="row">
+    <div class="col-8">
+      <div class="p-media-object u-no-margin--bottom">
+        <img src="https://assets.ubuntu.com/v1/402069bd-plex-media-server.png?w=90" class="p-media-object__image is-round" height="90" width="90">
```
Comment (anthonydillon) [inline templates/appliance/plex/index.html:15]: Needs an `alt`

File: templates/appliance/plex/index.html [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+        <div class="p-media-object__details">
+          <h1 class="p-heading--three u-no-margin--bottom">
+            Plex Media Server
+          </h1>
+          <p class="p-muted-heading"">Ubuntu Appliance</p>
```
Comment (anthonydillon) [inline templates/appliance/plex/index.html:20]: Double quotes still

File: templates/appliance/plex/index.html [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+
+<section class="p-strip--light">
+  <div class="row">
+    <div class="col-8">
+      <h2>Get started</h2>
+      <p>Once you have installed and set up your Plex Media Server Ubuntu appliance you will be taken through set up in your browser. If you have any more questions or want to go further, the developers of Plex Media Server have lots of documentation to help you get started.</p>
+      <a href="https://snapcraft.io/plexmediaserver" class="p-link--external">Get Started</a>
+    </div>
+    <div class="col-4 u-align--center u-vertically-center u-hide--small">
+      <img src="https://assets.ubuntu.com/v1/402069bd-plex-media-server.png?w=200" height="200" width="200">
```
Comment (anthonydillon) [inline templates/appliance/plex/index.html:139]: Missing alt

File: templates/appliance/plex/index.html [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+        <li class="p-list__item is-ticked">Cut the cord! Go premium with a Plex Pass and use Plex Live TV & DVR so you can watch and record free over-the-air HD TV, like NBC, ABC, CBS, and FOX, with any supported digital antenna and tuner</li>
+        <li class="p-list__item is-ticked">Enjoy free features, like remote access, sharing, security, and casting, see your media beautifully enhanced with rich descriptions, artwork, ratings, and more</li>
+        <li class="p-list__item is-ticked">Enjoy premium features with a Plex Pass subscription: DVR & Live TV, Camera Upload, Premium Photos, Mobile Sync, Parental Controls, VIP Perks, and much more</li>
+        <li class="p-list__item is-ticked">Want to use Plex on all of your devices, any time, anywhere? Visit <a href="https://www.plex.tv/" class="p-link--external">https://www.plex.tv/</a> for more information.</li>
+      </ul>
+    </div>
+    <div class="col-4">
+      <h2 class="p-heading--four">Snaps in the image</h2>
+      <div class="p-media-object">
+        <img src="https://assets.ubuntu.com/v1/402069bd-plex-media-server.png?w=75" class="p-media-object__image is-round" height="75" width="75">
```
Comment (anthonydillon) [inline templates/appliance/plex/index.html:90]: Missing alt

**Comment by lyubomir-popov** [comment]
Not sure if this is a known issue, but the slider doesn't work when I click the thumbnails, and the spacing top/bottom is not equally distributed within the grey strip:

**Comment by deadlight** [comment]
@lyubomir-popov I fixed that strip and the lightbox JS is being added in another branch

**Pull request merged**